### DISCUSSION TITLE
How to develop new decoder under Home Assistant OS

### DISCUSSION THREAD

**Discussion by TomyOne** [theengs/decoder#215] [New devices/things integration]
Hello,

I have installed Theengs Gateway as addon on Home Assistant OS:
https://gateway.theengs.io/install/install.html#install-theengs-gateway-as-an-add-on-in-home-assistant
It works great.

Now I would like to develop a new decoder.
The guides point me to this repo and they are clean I need to create new `*_json.h`.

What is not clear for me how do I get this new decoder into running instance of Theengs Gateway?

The Home Assistant OS running on rpbi4 is a bit tricky with all the docker images, but I have access to all files there
and cannot find there any decoder built library.

Guides are clear what to add where, but I am missing information how to build library and how to test my changes.

Thank you.

T.

**Comment by DigiH** [comment]
Hi @TomyOne,

Are you using Platformio to create and follow the [Adding decoders](https://decoder.theengs.io/participate/adding-decoders.html) instructions?

If so, then there you will also find the [test cases](https://github.com/theengs/decoder/blob/development/tests/BLE/test_ble.cpp) to which you can add test cases for your new decoder to build, run and verify in Platformio to check if everything is working fine.

These test cases are also run when submitting to your new decoder branch. So they should always be used as a first step to verify that a new decoder is running without issues and as desired, before implementing the new decoder into any project like Gateway.

You would need to fork Decoder, add you new decoder and also fork Theengs Gateway, update the Decoder submodule to your updated Decoder and then build and install (add-on in your case/docker/snap) and run it in your HA environment.

We are currently thinking about the possibility of creating an easier way to test new decoder branches in HA, but the above is currently the only procedure.

May I ask for which device you want to create this new decoder, as it might also be interesting to others and we could include it in the official Decoder release?

**Reply by TomyOne** [reply]
Hello @DigiH,

You can use Platformio for this? I only found platformio.ini file in some example.

But nevertheless sorry for the late reply sadly not much time for this currently.
And thank you for your reply and hints. :) 

I found the way to develop decoder for this and play with it how ti behaves.
When I have a time I will try it again.

But here is also guide what I did.

1. Clone repository of gateway
`git clone https://github.com/theengs/gateway.git`
`cd gateway/`
`git subomodule update --init --recursive`
2. Make changes as described in the [Adding decoders](https://decoder.theengs.io/participate/adding-decoders.html)
3. Build with cmake file in this repo.
`mkdir build && cd build`
`cmake ..`
`make`
4. This will compile compile `_decoder.cpython-310-x86_64-linux-gnu.so`
    * This file is then used by Python gateway repository.
    * So I copied this file into directory `https://github.com/theengs/gateway/tree/development/TheengsGateway` and run gateway as python module locally on PC with USB Bluetooth dongle.
5. Run gateway `python -m TheengsGateway -H "192.168.xx.xx" -u "username" -p "password" -pa 1 -D 1`
6. And test MQTT via MQTT Explorer

Its bit simplified I used Visual Studio Code to compile it with cmake addon.
You also have to add path to skbuild repository `https://github.com/scikit-build/scikit-build`
I added in in setting.json as
`"cmake.configureArgs": [`
`        "-DCMAKE_MODULE_PATH=/home/user/workspace/scikit-build/skbuild/resources/cmake/",`
`    ]`

So this maybe useful for someone later :).


Integration into the Home assistant I will get later after I fine tune decoder.

Its for the Asset Tag BLE from onsemi: https://www.onsemi.com/pub/collateral/um70019-d.pdf

**Reply by DigiH** [reply]
> You can use Platformio for this? I only found platformio.ini file in some example.

VSC I meant really ;) with CMake and the Test Explorer extensions Decoder can be built and the test suits can be run for decoder validation and regression testing.

> Its for the Asset Tag BLE from onsemi: https://www.onsemi.com/pub/collateral/um70019-d.pdf

Looks interesting! Will have to have a closer look at the sensors to see if they might be relevant for inclusion into Decoder release.

Thanks
